[clang][dataflow] Don't propagate result objects in unevaluated contexts

martinboehme opened the pull request:
Trying to do so can cause crashes -- see newly added test and the comments in
the fix.

We're starting to see a repeating pattern here: We're getting crashes because
`ResultObjectVisitor` and `getReferencedDecls()` don't agree on which parts of
the AST to visit and, hence, which fields should be modeled.

I think we should ensure consistency between these two parts of the code by
using a `RecursiveASTVisitor` in `getReferencedDecls()`[^1]; the
`Traverse...()` functions that control which parts of the AST we visit would go
in a common base class that would be used for both `ResultObjectVisitor` and
`getReferencedDecls()`.

I'd like to focus this PR, however, on a targeted fix for the current crash and
postpone the refactoring to a later PR (which will be easier to revert if there
are unintended side-effects).

[^1]: As an added bonus, this would make the code better structured and more
efficient than the current sequence of `if (dyn_cast<T>(...))` statements).

Xazax-hun commented on clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp:
```diff
@@ -350,6 +350,13 @@ class ResultObjectVisitor : public RecursiveASTVisitor<ResultObjectVisitor> {
     return RecursiveASTVisitor<ResultObjectVisitor>::TraverseDecl(D);
   }
 
+  bool TraverseDecltypeTypeLoc(DecltypeTypeLoc Node) {
```
There are other instances in the language that are unevaluated including: `typeid`, `noexcept`, `sizeof`. Do you anticipate similar problems with those?

martinboehme replied:
> There are other instances in the language that are unevaluated including: `typeid`, `noexcept`, `sizeof`. Do you anticipate similar problems with those?

Excellent point, thanks. I'll expand this PR to more generally exclude all unevaluated contexts. Will ping this PR when I've changed this.

Xazax-hun replied:
> It also turns out that, even though the operand of the noexcept operator in an unevaluated operand, it still shows up in the CFG for some reason

This surprises me, I can't think of a reason why we would want to have it in the CFG. This keyword appeards a lot in generic code, so maybe excluding it from the CFG might have measurable perf/memory benefits. I'd think it might be worth to give this a try and see if something breaks (in a separate PR).

martinboehme replied:
Makes sense, I'll take a look at that!

martinboehme commented:
Windows CI failure is in Driver/amdgpu-toolchain.c. This looks unrelated.

martinboehme committed:
[clang][dataflow] Don't propagate result objects in unevaluated contexts.

Trying to do so can cause crashes -- see newly added test and the comments in
the fix.

We're starting to see a repeating pattern here: We're getting crashes because
`ResultObjectVisitor` and `getReferencedDecls()` don't agree on which parts of
the AST to visit and, hence, which fields should be modeled.

I think we should ensure consistency between these two parts of the code by
using a `RecursiveASTVisitor` in `getReferencedDecls()`[^1]; the
`Traverse...()` functions that control which parts of the AST we visit would go
in a common base class that would be used for both `ResultObjectVisitor` and
`getReferencedDecls()`.

I'd like to focus this PR, however, on a targeted fix for the current crash and
postpone the refactoring to a later PR (which will be easier to revert if there
are unintended side-effects).

[^1]: As an added bonus, this would make the code better structured and more
efficient than the current sequence of `if (dyn_cast<T>(...))` statements).

Xazax-hun reviewed (approved):
LGTM, thanks!

cHackz18 commented:
Hello and good morning from the UK!

It looks as though this change caused a test failure on the following buildbot:
https://lab.llvm.org/buildbot/#/builders/216/builds/38446

Are you able to take a look see?

many thanks for yoru time,
Tom W

TomWeaver18 commented:
Reverted in 2252c5c42b95dd12dda0c7fb1b2811f943b21949

martinboehme commented:
> Hello and good morning from the UK!
> 
> It looks as though this change caused a test failure on the following buildbot: https://lab.llvm.org/buildbot/#/builders/216/builds/38446
> 
> Are you able to take a look see?
> 
> many thanks for yoru time, Tom W

Sorry, didn't see this in time.

> Reverted in [2252c5c](https://github.com/llvm/llvm-project/commit/2252c5c42b95dd12dda0c7fb1b2811f943b21949)

Thanks for the revert!

Will investigate and reland with a fix.
